File: experiments/source/ocaml/weighted_dot_pipeline.ml
open Loom

let[@loom.entry] weighted_dot_pipeline (weight : float) (x : Tensor1.t) (y : Tensor1.t) =
  let weighted_term xi yi =
    let prod = xi *. yi in
    weight *. prod
  in
  Tensor1.reduce_sum (Tensor1.map2 weighted_term x y)

File: experiments/source/python/weighted_dot_pipeline.py
import loom


@loom.entry
def weighted_dot_pipeline(weight: float, x: loom.Tensor1, y: loom.Tensor1) -> float:
    def weighted_term(xi: float, yi: float) -> float:
        prod = xi * yi
        return weight * prod

    return loom.Tensor1.reduce_sum(loom.Tensor1.map2(weighted_term, x, y))
